### ISSUE TITLE
What should a range of enums mean?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#8795]
Given an enum:

```chapel
enum color { red = 1, green = 3, blue = 2};
```

today, a range like `color.red..color.blue` results in the range `1..2` due to integer coercion (see issue #8794).  This is confusing both in that enums are not yielded by a loop over the range, and arguably in that `green` is omitted from the iteration even though it appears between `red` and `blue` in the enumeration.  This issue asks how such ranges of enums should be handled.  Some options might include:

* making them illegal (require casting to integer if one wants a range over the integer values)
* having them represent the symbols from the enum, respecting the order in which they were enumerated
* other?

We could also choose to restrict enum ranges to only be supported for enums with monotonically increasing values (or maybe even stride==1 enums?)

Note that we should probably resolve issue #8794 before worrying too much about this one.

**Comment by lydia-duncan** [comment]
I'd vote for having them represent the symbols from the enum, respecting the order in which they were enumerated

**Comment by bradcray** [comment]
> I'd vote for having them represent the symbols from the enum, respecting the order in which they were enumerated

I think that's natural, if they're supported.  A challenge to doing so is that if the integer values aren't stride-1 monotonic, the implementation of an iterator over such a range seems non-trivial / non-obvious to me.  For example, here's how I think the iterator over a range of colors might look:

```chapel
        enum color { red = 1, GREEN= 3, yellow = 2, black = 4 };

        // iterate over a range of colors
        iter range(color).these() {
          bool iterOn;

         if (this.unboundedLow()) then
          iterOn = true;

          if (low == red) then
            iterOn = true;
          if iterOn then
            yield red;
          if (high == red) then
            iterOn = false;

          if (low == GREEN) then
            iterOn = true;
          if iterOn then
            yield GREEN;
          if (high == GREEN) then
            iterOn = false;

          if (low == yellow) then
            iterOn = true;
          if iterOn then
            yield yellow;
          if (high == yellow) then
            iterOn = false;

          if (low == black) then
            iterOn = true;
          if iterOn then
            yield black;
          if (high == black) then
            iterOn = false;
        }
```

Then there's a question about how we'd make such an iterator work for strided ranges (maybe enum ranges just couldn't be strided...)

For a unit-strided enum, the iterator could be written more trivially as:

```chapel
        iter range(color).these() {
          const lowAsInt = low:int,
                highAsInt = high:int;
          for i in low..high do
            yield i: color;
        }
```

**Comment by bradcray** [comment]
The decision here was to have enum ranges respect the order in which the enum's symbols were declared.  This was implemented in PR #9820.
